packages/gatsby/index.d.ts

```diff
@@ -146,6 +146,35 @@ export type PageProps<
   serverData: ServerDataType
 }
 
+/**
+ * Props object passed into the [getServerData](https://www.gatsbyjs.com/docs/reference/rendering-options/server-side-rendering/) function.
+ */
+export type GetServerDataProps = {
+  headers: Map<string, unknown>
+  method: string
+  url: string
+  query?: Record<string, unknown>
+  params?: Record<string, unknown>
+  pageContext: Record<string, unknown>
+};
+
+/**
+ * The return type (promise payload) from the [getServerData](https://www.gatsbyjs.com/docs/reference/rendering-options/server-side-rendering/) function.
+ */
+export type GetServerDataReturn<ServerDataType = Record<string, unknown>> = {
+  headers?: Map<string, unknown>
+  props?: ServerDataType
+  status?: number
+};
+
+/**
+ * A shorthand type for combining the props and return type for the [getServerData](https://www.gatsbyjs.com/docs/reference/rendering-options/server-side-rendering/) function.
+ */
+export type GetServerData<ServerDataType> = (props: GetServerDataProps) => Promise<GetServerDataReturn<ServerDataType>>;
```

Comment:
(from our chat) - let's make `GetServerDataReturn` be a promise, so that users could do just:
```ts
export function getServerData(arg: GetServerDataProps): GetServerDataReturn {
 // ...
}
```

without having to wrap a thing in `Promise` themselves

Reply:
Addressed in [`ea9cf58` (#34406)](https://github.com/gatsbyjs/gatsby/pull/34406/commits/ea9cf582d608bae65e32abe0c79b33251e68552b)